src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -1015,6 +1015,35 @@ void ShenandoahHeapRegion::promote_in_place() {
   ShenandoahYoungGeneration* young_gen = heap->young_generation();
   size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
 
+  assert(get_top_before_promote() == tams, "Cannot promote regions in place if top has advanced beyond TAMS");
+
+  // Since this region may have served previously as OLD, it may hold obsolete object range info.
+  heap->card_scan()->reset_object_range(bottom(), end());
+  heap->card_scan()->mark_range_as_dirty(bottom(), get_top_before_promote() - bottom());
+
+  // TODO: use an existing coalesce-and-fill function rather than
+  // replicating the code here.
+  HeapWord* obj_addr = bottom();
+  while (obj_addr < tams) {
+    oop obj = cast_to_oop(obj_addr);
+    if (marking_context->is_marked(obj)) {
+      assert(obj->klass() != NULL, "klass should not be NULL");
+      // This thread is responsible for registering all objects in this region.  No need for lock.
+      heap->card_scan()->register_object_without_lock(obj_addr);
+      obj_addr += obj->size();
+    } else {
+      HeapWord* next_marked_obj = marking_context->get_next_marked_addr(obj_addr, tams);
+      assert(next_marked_obj <= tams, "next marked object cannot exceed tams");
+      size_t fill_size = next_marked_obj - obj_addr;
+      assert(fill_size >= ShenandoahHeap::min_fill_size(), "previously allocated objects known to be larger than min_size");
+      ShenandoahHeap::fill_with_object(obj_addr, fill_size);
+      heap->card_scan()->register_object_without_lock(obj_addr);
```

Comment:
On second thoughts, do we really need to register the dead objects? Does it help at all, since the preceding and succeeding live objects will be registered anyway. Not wrong, but I do wonder if it helps at all or not. Probably not a big deal either way since dead objects are likely very few in regions promoted in place.

Reply:
Existing code (and some existing assertions in ShenandoahVerifier) expect the filler objects to be registered.  You make a good point though.  We don't really need these to be registered.  If the first or only object in a card range is a filler, remembered set scanning can just ignore that.  The remembered set scanner wants to find the first non-filler object in the range.  We might want to address this in a separate ticket later, if we decide to keep the current remembered set representation longer term.

Reply:
Note that the cost of registering filler objects is minimal because there are never multiple contiguous filler objects residing within memory.  We always coalesce them together.  At most, the cost during remembered set scanning is that we look at one extra filler object per DIRTY card.

Reply:
I agree, I don't think it helps one way or other. In fact, the current card scanning code looks for the last object on the _previous_ card from which to start walking objects to scan. In the case that the filler object is entirely on a card, this does not make any difference. However, in the case that the filler object is straddling a card, registering it shortens the walk by one object, the one preceding the filler. I think any performance difference is a wash, and I'd just scratch my thought above.